src/GameOver.hs

```diff
@@ -0,0 +1,28 @@
+module GameOver(verifyGameOver) where
+import Data.Array
+import Game
+import Utils
+
+verifyGameOver :: Board -> State
+verifyGameOver board
+  | emptyCells == [] && noMergePossible = GameOver
+  | otherwise = Running
+  where
+    emptyCells = getEmptyCells board
+    noMergePossible = not (any id (map (\index -> verifyDirections index board) (indices board)))
+
+verifyDirections :: (Int, Int) -> Board -> Bool
+verifyDirections index board = 
+  verifyUp index board || verifyDown index board || verifyLeft index board || verifyRight index board
+
+verifyUp :: (Int, Int) -> Board -> Bool
+verifyUp (x, y) board = x > 0 && board ! (x, y) == board ! (x-1, y)
+
+verifyDown :: (Int, Int) -> Board -> Bool
+verifyDown (x, y) board = x < n-1 && board ! (x, y) == board ! (x+1, y)
```

Comment:
No instance for ‘Eq Cell’ arising from a use of ‘==’
• In the second argument of ‘(&&)’, namely
    ‘board ! (x, y) == board ! (x + 1, y)’
  In the expression:
    x < n - 1 && board ! (x, y) == board ! (x + 1, y)
  In an equation for ‘verifyDown’:
      verifyDown (x, y) board
        = x < n - 1 && board ! (x, y) == board ! (x + 1, y)typecheck(-Wdeferred-type-errors)